pkg/backup/file_sst_sink.go

```diff
@@ -331,11 +349,217 @@ func (s *fileSSTSink) write(ctx context.Context, resp exportedSpan) (roachpb.Key
+// if it does. This function should only be called after s.midRow has been updated for the last key/span written.
+func (s *fileSSTSink) maybeDoSizeFlush(ctx context.Context) error {
+	if s.shouldDoSizeFlush() {
+		return s.doSizeFlush(ctx)
+	} else {
+		log.VEventf(ctx, 3, "continuing to write to backup file %s of size %d", s.outName, s.flushedSize)
+		return nil
+	}
+}
+
+// maybeDoSizeFlushWithNextKey is identical to maybeDoSizeFlush, except information about the next key to be written
+// is provided. This is used to determine if the next key can act as an exclusive end key for the last
+// BackupManifest_File. The updated span of the next key is returned, as it may be set to contiguously extend the last
+// span. This function should only be called after s.midRow has been updated for the last key/span
+// written.
+func (s *fileSSTSink) maybeDoSizeFlushWithNextKey(
+	ctx context.Context, nextKey roachpb.Key, nextSpan roachpb.Span, startTime, endTime hlc.Timestamp,
+) (roachpb.Span, error) {
+	if len(s.flushedFiles) == 0 || !s.shouldDoSizeFlush() {
+		log.VEventf(ctx, 3, "continuing to write to backup file %s of size %d", s.outName, s.flushedSize)
+		return nextSpan, nil
+	}
+	// If the span of the next key to be written covers the same span as the last BackupManifest_File, then the next key
+	// can act as an exclusive end key for the last file's span. Additionally, the span of the nextKey will be updated
+	// so that it contiguously extends with the last span.
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if sameAsFileSpan(nextSpan, startTime, endTime, lastFile) {
+		lastFile.Span.EndKey = nextKey
+		nextSpan.Key = nextKey
+	}
+	return nextSpan, s.doSizeFlush(ctx)
+}
+
+// shouldDoSizeFlush returns true if the current backup file exceeds the targetFileSize and we are currently not mid-row.
+// Should only be called after s.midRow has been updated for the last key/span written.
+func (s *fileSSTSink) shouldDoSizeFlush() bool {
+	return s.flushedSize > targetFileSize.Get(s.conf.settings) && !s.midRow
+}
+
+// doSizeFlush flushes the buffered files with size exceeded as the reason.
+func (s *fileSSTSink) doSizeFlush(ctx context.Context) error {
+	s.stats.sizeFlushes++
+	log.VEventf(ctx, 2, "flushing backup file %s with size %d", s.outName, s.flushedSize)
+	return s.flushFile(ctx)
+}
+
+// extendsLastSpan returns true if the new span can be added to the last span in the flushedFiles slice.
+// If the last written key was mid-row, the new span is always considered to extend the last span.
+// If the new span is a contiguous extension of the last span, it is also considered to extend the last span.
+// extendSameSpan determines if an identical span should be considered an extension of the last span.
+// This should only be called after s.midRow has been updated for the last key/span written.
+func (s *fileSSTSink) shouldExtendLastFile(
```

Comment:
Nit: might just make this a regular non-receiver function like the rest of them below.